test/unit/puppetlabs/jruby_utils/lockable_pool_test.clj

```diff
@@ -584,7 +584,24 @@
     (let [pool (create-populated-pool 1)
           pill (str "i'm a pill")]
       (.insertPill pool pill)
-      (is (identical? (.borrowItem pool) pill)))))
+      (is (identical? (.borrowItem pool) pill))
+      (testing "subsequent borrows return the same pill"
+        (is (identical? (.borrowItem pool) pill)))))
+  (testing "when borrow is blocked, inserting a pill unblocks it"
+    (let [pool (create-populated-pool 1)
+          pill (str "I'm just a pill, yes I'm only a pill")
+          instance (.borrowItem pool)
+          borrow-promise (promise)
+          blocked-borrow (future (let [instance (.borrowItem pool)]
+                                   (deliver borrow-promise instance)))]
+      (is (= 0 (.size pool)))
+      ; Pool is empty and the borrow future is blocked
+      (is (not (realized? borrow-promise)))
```

Comment:
Seems like you could just return the borrowed instance from the `future` and check here that the `blocked-borrow` future has not been realized rather than creating a separate promise?  Sometimes we will use a promise in a future, though, so that we can ensure that we at least got into the future before we make any assertions, though, like this:

```clj
about-to-borrow-promise (promise)
blocked-borrow (future (do (deliver about-to-borrow-promise true) (.borrowItem pool)))
...
(is @about-to-borrow-promise)
(is (not (realized? blocked-borrow)))
(.insertPill pool pill)
(let [borrow-result (timed-deref blocked-borrow)]
  (is (identical? borrow-result pill)))
```